ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifierImpl.java

```diff
@@ -155,36 +159,46 @@ private SafeFuture<Optional<ExecutionPayloadContext>> internalGetPayloadId(
 
       LOG.warn(
           "No suitable payloadId for block production at slot {}, requesting a new one to the EL",
           blockSlot);
 
       // to make sure that we deal with the same data when calculatePayloadAttributes asynchronously
       // returns, we save locally the current class reference.
-      ForkChoiceUpdateData localForkChoiceUpdateData = forkChoiceUpdateData;
-      return proposersDataManager
-          .calculatePayloadBuildingAttributes(blockSlot, inSync, localForkChoiceUpdateData, true)
-          .thenCompose(
-              newPayloadAttributes -> {
-
-                // we make the updated local data global, reverting any potential data not yet sent
-                // to EL
+      final ForkChoiceUpdateData localForkChoiceUpdateData = forkChoiceUpdateData;
+
+      return forkChoiceStateProvider
+          .getForkChoiceState()
+          .thenCombine(
+              proposersDataManager.calculatePayloadBuildingAttributes(
+                  blockSlot, inSync, localForkChoiceUpdateData, true),
+              (forkChoiceState, payloadBuildingAttributes) -> {
                 forkChoiceUpdateData =
-                    localForkChoiceUpdateData.withPayloadBuildingAttributes(newPayloadAttributes);
+                    localForkChoiceUpdateData
+                        .withForkChoiceState(forkChoiceState)
+                        .withPayloadBuildingAttributes(payloadBuildingAttributes);
+
                 sendForkChoiceUpdated();
-                return forkChoiceUpdateData
-                    .getExecutionPayloadContext()
-                    .thenApply(
-                        executionPayloadContext -> {
-                          if (executionPayloadContext.isEmpty()) {
-                            throw new IllegalStateException(
-                                "Unable to obtain an executionPayloadContext");
-                          }
-                          return executionPayloadContext;
-                        });
-              });
+
+                if (!forkChoiceUpdateData.isPayloadIdSuitable(parentExecutionHash, timestamp)) {
+                  throw new IllegalStateException(
+                      "payloadId still not suitable after requesting a new one via FcU with recalculated data");
+                }
+
+                return forkChoiceUpdateData.getExecutionPayloadContext();
+              })
+          .thenCompose(
+              executionPayloadContextFuture ->
+                  executionPayloadContextFuture.thenApply(
+                      maybeExecutionPayloadContext -> {
+                        if (maybeExecutionPayloadContext.isEmpty()) {
+                          throw new IllegalStateException(
+                              "Unable to obtain an executionPayloadContext");
+                        }
+                        return maybeExecutionPayloadContext;
+                      }));
```

Comment:
We can avoid suppressing ignored futures with:
```java
      return forkChoiceStateProvider
          .getForkChoiceState()
          .thenCombine(
              proposersDataManager.calculatePayloadBuildingAttributes(
                  blockSlot, inSync, localForkChoiceUpdateData, true),
              (forkChoiceState, payloadBuildingAttributes) -> {
                forkChoiceUpdateData =
                    localForkChoiceUpdateData
                        .withForkChoiceState(forkChoiceState)
                        .withPayloadBuildingAttributes(payloadBuildingAttributes);

                sendForkChoiceUpdated();

                if (!forkChoiceUpdateData.isPayloadIdSuitable(parentExecutionHash, timestamp)) {
                  throw new IllegalStateException(
                      "payloadId still not suitable after requesting a new one via FcU with recalculated data");
                }
                return forkChoiceUpdateData;
              })
          .thenCompose(ForkChoiceUpdateData::getExecutionPayloadContext)
          .thenApply(
              maybeExecutionPayloadContext -> {
                if (maybeExecutionPayloadContext.isEmpty()) {
                  throw new IllegalStateException("Unable to obtain an executionPayloadContext");
                }
                return maybeExecutionPayloadContext;
              });
```

Reply:
ty :)